Show which case panels affected genes belong to in Clinical SNV and INDELs variant pages for variants that affect more than one gene

alkc opened the pull request:
## Gene panel info in summary table:

This PR closes #4233 by adding a hovertip with additional info for listed gene panels in SNV indel variant view. 
The hovertip is activated only when the variant covers more than one gene and lists which of the multiple genes are included in the listed panels:

![hovertip-update](https://github.com/Clinical-Genomics/scout/assets/13668555/9b9d086e-0e85-4a29-8b47-afd80927b041)

The variant genes and case panels are associated with each other after `update_case_panels()` `update_variant_case_panels()` have been called in the variant controller. The matching panels are added to each gene by a new function `associate_variant_genes_with_case_panels()`, which modifies each variant gene object to list the case panels it belongs to.

## Other changes

I got carried away when working on this and changed a few other things:

### Projections

I've enabled projections in `PanelHandler.panel()` and `PanelHandler.gene_panel()` getters to be able to instruct scout to only return the gene lists in `PanelHandler.panel_to_genes()`.

### Tests

Added tests that cover `PanelHandler.panel_to_genes()`, since I became nervous when I  was poking around in that function.

### Documentation and Type annotations

I've updated function docstrings and added py3.8 type hints in the parts of the code that are affected by this update. Could continue this work in separate updates, if interesting.

<details>
<summary>Testing on cg-vm1 server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. Make sure the PR is pushed and available on [Docker Hub](https://hub.docker.com/repository/docker/clinicalgenomics/scout-server-stage)
1. Fist book your testing time using the Pax software available at [https://pax.scilifelab.se/](https://pax.scilifelab.se). The resource you are going to call dibs on is `scout-stage` and the server is `cg-vm1`.
1. `ssh <USER.NAME>@cg-vm1.scilifelab.se`
1. `sudo -iu hiseq.clinical`
1. `ssh localhost`
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `podman ps`
1. Stop the service with current deployed branch: `systemctl --user stop scout.target`
1. Start the scout service with the branch to test: `systemctl --user start scout@<this_branch>`
1. Make sure the branch is deployed: `systemctl --user status scout.target`
1. After testing is done, repeat procedure at [https://pax.scilifelab.se/](https://pax.scilifelab.se), which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>

<details>
<summary>Testing on hasta server (Clinical Genomics Stockholm)</summary>

**Prepare for testing**
1. `ssh <USER.NAME>@hasta.scilifelab.se`
1. Book your testing time using the Pax software. `us; paxa -u <user> -s hasta -r scout-stage`. You can also use the WSGI Pax app available at [https://pax.scilifelab.se/](https://pax.scilifelab.se).
1. (optional) Find out which scout branch is currently deployed on cg-vm1: `conda activate S_scout; pip freeze | grep scout-browser`
1. Deploy the branch to test: `bash /home/proj/production/servers/resources/hasta.scilifelab.se/update-tool-stage.sh -e S_scout -t scout -b <this_branch>`
1. Make sure the branch is deployed: `us; scout --version`
1. After testing is done, repeat the `paxa` procedure, which will release the allocated resource (`scout-stage`) to be used for testing by other users.
</details>


**How to test**:
1. Set up this update in an env containing cases with variants that cover multiple genes
2. Ensure that that hovertip is activated when accessing variants covering multiple genes
3. Ensure that the the case panel info correctly lists genes that are included in the panel
4. Ensure that the hovertip is inactivated when displaying variant affecting only one gene

**Expected outcome**:
The functionality should be working
Take a screenshot and attach or copy/paste the output.

**Review:**
- [ ] code approved by
- [ ] tests executed by

northwestwitch commented on CHANGELOG.md:
```diff
@@ -14,12 +14,15 @@ About changelog [here](https://keepachangelog.com/en/1.0.0/)
 
 ## [4.74]
 ### Added
+- Hovertip to gene panel names with associated genes in variant view, when variant covers more than one gene 
```
Great description of the changes! 🥇 

Since we have released a new version of scout yesterday (4.74.1] could you move these lined in a new release item, like this:

```
## [unreleased]
### Added
### Changed
```

northwestwitch commented on scout/adapter/mongo/panel.py:
```diff
@@ -195,21 +196,25 @@ def add_gene_panel(self, panel_obj, replace=False):
 
-    def panel(self, panel_id):
+    def panel(
+        self, panel_id: Union[str, ObjectId], projection: Optional[Dict] = None
```
💯

northwestwitch commented on scout/adapter/mongo/panel.py:
```diff
@@ -195,21 +196,25 @@ def add_gene_panel(self, panel_obj, replace=False):
+    ) -> Optional[Dict]:
         """Fetch a gene panel by '_id'.
 
         Args:
```
Now that you defined the typehints I think it is fine to remove both Args and Returns.

northwestwitch commented on scout/adapter/mongo/panel.py:
```diff
@@ -227,25 +232,26 @@ def delete_panel(self, panel_obj):
 
-        If no panel is sent return all panels
-
         Args:
```
Also here, now it is safe to remove Args and Returns

northwestwitch commented on scout/adapter/mongo/panel.py:
```diff
@@ -345,23 +351,29 @@ def panels_to_genes(
         panel_obj = None
         if panel_id:
-            panel_obj = self.panel(panel_id)
+            panel_obj = self.panel(panel_id, genes_projection)
```
```suggestion
            panel_obj = self.panel(panel_id, projection=genes_projection)
```

northwestwitch commented on tests/adapter/mongo/test_panel_handler.py:
```diff
@@ -1,6 +1,7 @@
 from pprint import pprint as pp
```
Would be nice to remove the unused pp import as flagged by the GitHub action

northwestwitch reviewed (approved):
Works perfectly! It's a useful feature that will be appreciated by clinicians. Great with the projection! 💯

alkc commented:
Thanks for the review and input @northwestwitch! 

I've cleaned up the docstrings where possible, but left some of the arg/return descriptions where I thought it would be useful for future code readers. 

I've also removed the unused import from the panel tests + some other unused vars from other test functions.

If the recent changes look OK, then I'll merge this PR after lunch!

northwestwitch commented:
> Thanks for the review and input @northwestwitch!
> 
> I've cleaned up the docstrings where possible, but left some of the arg/return descriptions where I thought it would be useful for future code readers.
> 
> I've also removed the unused import from the panel tests + some other unused vars from other test functions.
> 
> If the recent changes look OK, then I'll merge this PR after lunch!

Thank you for the PR! Feel free to merge it whenever you are ready
